diff --ed a/config/usage.cfg b/config/usage.cfg
--- a/config/usage.cfg
+++ b/config/usage.cfg
330a
setdesc "thirdpersonzoom" "if set and thirdpersoncursor is 0 then zoom happens in third person intead of first person" "<value>"
.
diff --git a/src/game/game.cpp b/src/game/game.cpp
index 1ba4455..93162f7 100644
--- a/src/game/game.cpp
+++ b/src/game/game.cpp
@@ -76,10 +76,11 @@ namespace game
     VAR(IDF_PERSIST, thirdpersoninterp, 0, 100, VAR_MAX);
     FVAR(IDF_PERSIST, thirdpersondist, FVAR_NONZERO, 14, 20);
     FVAR(IDF_PERSIST, thirdpersonside, -20, 7, 20);
-    VAR(IDF_PERSIST, thirdpersoncursor, 0, 1, 2);
+    VAR(IDF_PERSIST, thirdpersoncursor, 0, 0, 2);
     FVAR(IDF_PERSIST, thirdpersoncursorx, 0, 0.5f, 1);
     FVAR(IDF_PERSIST, thirdpersoncursory, 0, 0.5f, 1);
     FVAR(IDF_PERSIST, thirdpersoncursordist, 0, 256, FVAR_MAX);
+    VAR(IDF_PERSIST, thirdpersonzoom, 0, 1, 1); // Works only with thirdpersoncursor=0; 0 = switch to 1p view for zooming in 3p, 1 = stay in 3p view for zooming in 3p
 
     VARF(0, follow, -1, -1, VAR_MAX, followswitch(0));
 
@@ -453,7 +454,7 @@ namespace game
         if(!viewonly && (d->state == CS_DEAD || d->state == CS_WAITING)) return true;
         if(player1->state == CS_EDITING) return false;
         if(player1->state >= CS_SPECTATOR && d == player1) return false;
-        if(d == focus && inzoom()) return false;
+        if(d == focus && inzoom() && (!thirdpersonzoom || thirdpersoncursor > 0)) return false; // Change view to 1p unless thirdpersonzoom
         if(!(d != player1 ? followthirdperson : thirdperson)) return false;
         return true;
     }
@@ -3190,6 +3191,7 @@ namespace game
     void renderplayer(gameent *d, int third, float trans, float size, bool early = false)
     {
         if(d->state == CS_SPECTATOR) return;
+        if(d == player1 && thirdpersonview() && inzoom()) return; // Hide player model while zooming in 3p
         if(trans <= 0.f || (d == focus && !(third == 1 ? thirdpersonmodel : firstpersonmodel)))
         {
             if(d->state == CS_ALIVE && rendernormally && (early || d != focus))
