diff --git a/src/game/server.cpp b/src/game/server.cpp
index 63d3de2..219acce 100644
--- a/src/game/server.cpp
+++ b/src/game/server.cpp
@@ -7012,8 +7012,6 @@ namespace server
                         hasgameinfo = true;
                         sents[n].o = vec(o).div(DMF);
                         packetbuf q(MAXTRANS, ENET_PACKET_FLAG_RELIABLE);
-                        putint(q, N_CLIENT);
-                        putint(q, ci->clientnum);
                         putint(q, N_EDITENT);
                         putint(q, n);
                         putint(q, o.x);
@@ -7022,7 +7020,12 @@ namespace server
                         putint(q, sents[n].type);
                         putint(q, sents[n].attrs.length());
                         loopvk(sents[n].attrs) putint(q, sents[n].attrs[k]);
-                        sendpacket(-1, 1, q.finalize(), ci->clientnum);
+                        packetbuf cep(MAXTRANS, ENET_PACKET_FLAG_RELIABLE);
+                        putint(cep, N_CLIENT);
+                        putint(cep, ci->clientnum);
+                        putuint(cep, q.length());
+                        cep.put(q.getbuf(), q.length());
+                        sendpacket(-1, 1, cep.finalize(), ci->clientnum);
                         if(tweaked)
                         {
                             if(enttype[sents[n].type].usetype == EU_ITEM) setspawn(n, true, true, true);
