diff --git a/config/menus/servers.cfg b/config/menus/servers.cfg
index 181009a..645d1a1 100644
--- a/config/menus/servers.cfg
+++ b/config/menus/servers.cfg
@@ -10,7 +10,7 @@ sinfoservers = 0
 sinfonumsrv = 0
 sinfouitime = 0
 
-sinfotypes = [ "" "status" "name" "port" "qport" "desc" "mode" "muts" "map" "time" "players" "maxplayers" "ping" ]
+sinfotypes = [ "" "status" "name" "port" "qport" "desc" "mode" "muts" "map" "time" "players" "maxplayers" "ping" "priority" ]
 sinfomodify = [
     sinfotlist = $arg1
     sinfotabsl = (? (> $sinfotlist 0) $sinfotlist (- 0 $sinfotlist))
@@ -74,6 +74,7 @@ servermenu = [
         sinfohandle = (getserver $i 0 8)
         sinfoflags = (getserver $i 0 9)
         sinfobran = (getserver $i 0 10)
+        sinfoprio = (getserver $i 0 11)
         sinfostatsflag = 0
         loop fi (stringlen $sinfoflags) [
             if (=s (substring $sinfoflags $fi 1) "s") [ sinfostatsflag = 1 ]
@@ -193,10 +194,13 @@ servermenu = [
                 guilist [
                     guibutton (format "^fw%1" $sinfodesc)
                     guifont "little" [ guicenter [ guibutton (format "^fd(^fa%1^fd)" $sinfonpid) ] ]
-                    guifont "little" [ guicenter [ guibutton (format "^fd[^fc%1^fd]" $sinfohandle) ] ]
+                    if $sinfohandle [
+                        guifont "little" [ guicenter [ guibutton (format "^fd[^fc%1^fd]" $sinfohandle) ] ]
+                    ]
                     if $sinfostatsflag [
                         guifont "little" [ guicenter [ guibutton "^fd[^fcstatistics^fd]" ] ]
                     ]
+                    guifont "little" [ guicenter [ guibutton (format "^fd[^fcprio %1^fd]" $sinfoprio) ] ]
                 ]
                 guispring 1
                 if $sinfoactive [
diff --git a/src/game/client.cpp b/src/game/client.cpp
index d3a7669..616559c 100644
--- a/src/game/client.cpp
+++ b/src/game/client.cpp
@@ -3242,19 +3242,9 @@ namespace client
     {
         int ac = 0, bc = 0;
         if(a->address.host == ENET_HOST_ANY || a->ping >= serverinfo::WAITING || a->attr.empty()) ac = -1;
-        else
-        {
-            ac = a->attr[0] == VERSION_GAME ? 0x7FFF : clamp(a->attr[0], 0, 0x7FFF-1);
-            ac <<= 16;
-            ac |= clamp(1 + a->priority, 1, 0xFFFF);
-        }
+        else ac = a->attr[0] == VERSION_GAME ? 0x7FFF : clamp(a->attr[0], 0, 0x7FFF-1);
         if(b->address.host == ENET_HOST_ANY || b->ping >= serverinfo::WAITING || b->attr.empty()) bc = -1;
-        else
-        {
-            bc = b->attr[0] == VERSION_GAME ? 0x7FFF : clamp(b->attr[0], 0, 0x7FFF-1);
-            bc <<= 16;
-            bc |= clamp(1 + b->priority, 1, 0xFFFF);
-        }
+        else bc = b->attr[0] == VERSION_GAME ? 0x7FFF : clamp(b->attr[0], 0, 0x7FFF-1);
         if(ac > bc) return -1;
         if(ac < bc) return 1;
 
@@ -3350,6 +3340,11 @@ namespace client
                     retsw(aa->ping, ab->ping, true);
                     break;
                 }
+                case SINFO_PRIO:
+                {
+                    retsw(aa->priority, ab->priority, false);
+                    break;
+                }
                 default: break;
             }
         }
@@ -3399,6 +3394,7 @@ namespace client
                         case 8: result(si->authhandle); break;
                         case 9: result(si->flags); break;
                         case 10: result(si->branch); break;
+                        case 11: intret(si->priority); break;
                     }
                     break;
                 case 1:
diff --git a/src/game/game.h b/src/game/game.h
index 537d10c..4b4ff6c 100644
--- a/src/game/game.h
+++ b/src/game/game.h
@@ -207,7 +207,7 @@ enum {
 #define MM_OPENSERV (MM_AUTOAPPROVE|(1<<MM_OPEN))
 
 enum { MM_OPEN = 0, MM_VETO, MM_LOCKED, MM_PRIVATE, MM_PASSWORD };
-enum { SINFO_NONE = 0, SINFO_STATUS, SINFO_NAME, SINFO_PORT, SINFO_QPORT, SINFO_DESC, SINFO_MODE, SINFO_MUTS, SINFO_MAP, SINFO_TIME, SINFO_NUMPLRS, SINFO_MAXPLRS, SINFO_PING, SINFO_MAX };
+enum { SINFO_NONE = 0, SINFO_STATUS, SINFO_NAME, SINFO_PORT, SINFO_QPORT, SINFO_DESC, SINFO_MODE, SINFO_MUTS, SINFO_MAP, SINFO_TIME, SINFO_NUMPLRS, SINFO_MAXPLRS, SINFO_PING, SINFO_PRIO, SINFO_MAX };
 enum { SSTAT_OPEN = 0, SSTAT_LOCKED, SSTAT_PRIVATE, SSTAT_FULL, SSTAT_UNKNOWN, SSTAT_MAX };
 
 enum {
